File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/fletcher16.cpp
#include "freight/hashes.h"

namespace freight {

// fletcher16 over raw bytes.
uint64_t hash_fletcher16(const std::string& data) {
  uint32_t low = 0u;
  uint32_t high = 0u;
  for (size_t i = 0; i < data.size(); ++i) {
    low = (low + static_cast<uint8_t>(data[i])) % 255u;
    high = (high + low) % 255u;
  }
  return static_cast<uint64_t>((high << 8) | low);
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Fletcher16.java
package com.freight.hash;

/** fletcher16 over raw bytes. */
public final class Fletcher16 implements HashAlgorithm {

    @Override
    public String name() {
        return "fletcher16";
    }

    @Override
    public long apply(byte[] data) {
        long low = 0L;
        long high = 0L;
        for (int i = 0; i < data.length; i++) {
            low = (low + (data[i] & 0xFF)) % 255L;
            high = (high + low) % 255L;
        }
        return ((high << 8) | low) & 0xFFFFL;
    }
}
